src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,321 @@
+
+@[simp]
+lemma rel_iff (H K : subgroup G) (x y : G) :
+  double_coset_rel H K x y  ↔  (∃ (a ∈ H) (b ∈ K), y = a * x * b) := iff.rfl
+
+lemma left_bot_eq_left_group_rel (H : subgroup G) :
+  (double_coset_rel ⊥ H) = (quotient_group.left_rel H).rel :=
+begin
+  show (double_coset_rel ⊥ H) = (quotient_group.left_rel H).r,
+  ext,
+  simp only [exists_prop, one_mul, subgroup.mem_bot, exists_eq_left, rel_iff],
+  split,
+  rintro ⟨a, ha⟩,
+  rw [quotient_group.left_rel, ha.2],
+  simp only [ha.left, inv_mul_cancel_left],
+  rw quotient_group.left_rel,
+  dsimp only,
+  intro h,
+  use (⟨x⁻¹ * x_1, h⟩ : H),
+  simp only [h, mul_inv_cancel_left, eq_self_iff_true, and_self, subgroup.coe_mk],
+end
+
+
+lemma right_bot_eq_right_group_rel (H : subgroup G) :
+  (double_coset_rel H ⊥) = (quotient_group.right_rel H).rel :=
+begin
+  show (double_coset_rel H ⊥) = (quotient_group.right_rel H).r,
+  ext,
+  simp only [exists_prop, one_mul, subgroup.mem_bot, exists_eq_left, rel_iff],
+  split,
+  rintro ⟨a, ha⟩,
+  rw [quotient_group.right_rel, ha.2],
+  simp only [ha.left, mul_one, mul_inv_cancel_right],
+  rw quotient_group.right_rel,
+  dsimp only,
+  intro h,
+  use (⟨x_1 * x⁻¹, h⟩ : H),
+  simp only [h, mul_one, eq_self_iff_true, and_self, subgroup.coe_mk, inv_mul_cancel_right],
+end
+
+lemma rel_reflex {H K : subgroup G} : reflexive (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  simp only [exists_prop],
+  intro x,
+  use 1,
+  simp only [H.one_mem, K.one_mem, one_mul, exists_eq_right, self_eq_mul_right, and_self],
+end
+
+lemma rel_symm {H K : subgroup G} : symmetric (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  intros x y,
+  simp only [and_imp, exists_prop, forall_exists_index],
+  intros a ha b hb hx,
+  use a⁻¹,
+  have haa := subgroup.inv_mem H ha,
+  simp only [haa, true_and],
+  use b⁻¹,
+  have hbb := subgroup.inv_mem K hb,
+  simp only [hbb, true_and],
+  rw hx,
+  simp_rw ← mul_assoc,
+  simp only [one_mul, mul_left_inv, mul_inv_cancel_right],
+end
+
+lemma rel_trans {H K : subgroup G} : transitive (double_coset_rel H K) :=
+begin
+  rw double_coset_rel,
+  intros x y z,
+  simp only [and_imp, exists_prop, forall_exists_index],
+  intros a ha b hb hxy c hc d hd hyz,
+  use c * a,
+  have hac := H.mul_mem hc ha,
+  simp only [hac, true_and],
+  use b*d,
+  have hdb :=  (K.mul_mem hb hd),
+  simp only [hdb, true_and, hyz, hxy, ← mul_assoc],
+end
+
+lemma rel_is_equiv  {H K : subgroup G} : equivalence (double_coset_rel H K) :=
+⟨rel_reflex, rel_symm, rel_trans⟩
+
+/--The setoid defined by the double_coset relation-/
+def setoid (H K : subgroup G) : setoid G :=
+⟨double_coset_rel H K, rel_is_equiv⟩
+
+/--Quotient of `G` by the double coset relation, i.e. `H \ G / K`-/
+def quotient (H K : subgroup G) : Type* := quotient (setoid H K)
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def doset (s t : set α) (a : α) : set α := { b : α | ∃ x : (s × t), b = x.1 * a * x.2 }
```

Comment:
I wonder if it's possible to take advantage of the existing `set.mul`. For example, this has the potential to make `sub_doset` nicer: `H * b * K = H * h * a * k * K = H * a * K`.
```suggestion
open_locale pointwise

/--The double_coset as an element of `set α` corresponding to `s a t` -/
def doset (s t : set α) (a : α) : set α := s * {a} * t
```

Reply:
So I've just tryed this. I'm probably not understanding what you mean, but what I end up with is something like `H * b * K = H * {h * a * k }* K ` but it seems I annoying to then remove the `{}` unless there is something simple I'm missing.

Reply:
Maybe we're missing `{x} * {y} = {x * y}`? Then you could try something like this: `H * {b} * K = H * {h * a * k} * K = H * {h} * a * {k} * K = H * a * K`.

Reply:
Here's a start. Seems like you might need a lemma about how pointwise multiplication interacts with subgroups.
```
/--The double_coset as an element of `set α` corresponding to `s a t` -/
def doset (s t : set α) (a : α) : set α := s * {a} * t

@[simp]
lemma doset_mem (s t : set α) (a b : α) : b ∈ (doset s t a) ↔ ∃ x : (s × t), b = x.1 * a * x.2 :=
⟨λ ⟨_, y, ⟨x, _, hx, rfl, rfl⟩, hy, h⟩, ⟨⟨⟨x, hx⟩, ⟨y, hy⟩⟩, h.symm⟩,
  λ ⟨⟨x, y⟩, h⟩, ⟨x * a, y, ⟨x, a, x.2, rfl, rfl⟩, y.2, h.symm⟩⟩

lemma sub_doset  (H K : subgroup G) (a b : G) (hb : b ∈ doset H.1 K a) :
  (doset H.1 K b) = (doset H K a) :=
begin
  obtain ⟨_, k, ⟨h, a, hh, (rfl : _ = _), rfl⟩, hk, rfl⟩ := hb,
  rw [doset, doset, ←set.singleton_mul_singleton, ←set.singleton_mul_singleton],
  rw [mul_assoc, mul_assoc, ←mul_assoc, ←mul_assoc H.carrier],
  have key1 : H.carrier * {h} = H := sorry,
  have key2 : ({k} : set G) * K = K := sorry,
  rw [key1, key2],
end
```